### PR TITLE
Add BigEarthNet Version2

### PR DESCRIPTION
**Pull request by nilsleh**
Superseeds #2371 after talking to Ando. 

After taking a look, the new version comes with a `metadata.parquet` file, which makes data handling quiet a bit more straightforward. With a `version=2` argument, I felt like there would be many nested if statements and therefore, cleaner to do it this way. If there was a similar `metadata.parquet` file for V1, then this could be made more condensed.

example_benv2

### CODE REVIEW

File: torchgeo/datasets/bigearthnet.py [hunk 1]
```diff
@@ -495,7 +504,7 @@ def _download(self, url: str, filename: Path, md5: str) -> None:
             filename: output filename to write downloaded file
             md5: md5 of downloaded file
         """
-        if not os.path.exists(filename):
+        if not os.path.exists(os.path.join(self.root, filename)):
```
Comment (nilsleh) [inline torchgeo/datasets/bigearthnet.py:507]: Saw this, and actually I think this is required, because it should check whether the file exists in root already right?

File: torchgeo/datasets/bigearthnet.py [hunk 2]
```diff
@@ -573,3 +582,325 @@ def plot(
...
+    def _load_target(self, index: int) -> Tensor:
+        """Load the target mask for a single image.
+
+        Args:
+            index: index to return
+
+        Returns:
+            the target label
+        """
+        indices = self.metadata_df.iloc[index]['labels']
```
Comment (nilsleh) [inline torchgeo/datasets/bigearthnet.py:792]: Also need to check the labels. In the V1 class it only seems possible to select the 19 label versions, because with selecting 43 they also get mapped to 19 if I understand correctly.

File: torchgeo/datasets/bigearthnet.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datasets/bigearthnet.py:507]: Good catch. Actually, this line isn't necessary because we already check if the zipfile exists and extract it before we reach the download. Can you open a separate PR to fix this so we can backport it to 0.6.3?

**Comment by nilsleh** [comment]
@ando-shah in case you wanna have a look, since you have experience with the dataset already and find anything.

File: torchgeo/datasets/bigearthnet.py [hunk 3]
```diff
@@ -573,3 +585,441 @@ def plot(
         if suptitle is not None:
             plt.suptitle(suptitle)
         return fig
+
+
+class BigEarthNetV2(NonGeoDataset):
```
Comment (adamjstewart) [inline torchgeo/datasets/bigearthnet.py:590]: Could also subclass BigEarthNet if you want less code duplication. Or make a shared base class. Depends on how similar they are.

File: torchgeo/datasets/bigearthnet.py [hunk 4]
```diff
@@ -573,3 +585,441 @@ def plot(
...
+        else:
+            paths = sorted(paths)
+
+        images = []
+        for path in paths:
+            with rasterio.open(path) as dataset:
+                array = dataset.read(
+                    indexes=1,
+                    out_shape=self.image_size,
+                    out_dtype='int32',
```
Comment (adamjstewart) [inline torchgeo/datasets/bigearthnet.py:820]: Should we load as float32 since we're later converting to float32 anyway? Or just load in the original dtype?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Once this is done I would love to squash this on top of #2371 so that @ando-shah gets some credit.

File: requirements/datasets.txt [hunk 5]
```diff
@@ -6,3 +6,4 @@ pandas[parquet]==2.2.3
 pycocotools==2.0.8
 scikit-image==0.25.0
 scipy==1.15.1
+zstandard==0.23.0
```
[the remark this answers is not in the document]
Comment (nilsleh) [reply requirements/datasets.txt:9]: mhm okay, this is a lot of work though

[the remark this answers is not in the document]
Comment (adamjstewart) [reply requirements/datasets.txt:9]: Is it a lot of work? You already have the data downloaded and extracted. You just need to compress, upload, and change some URLs. Seems like less work than adding this to `pyproject.toml` and finding the minimum supported version.

We only have a single dataset that needs this dep. We already removed our zipfile-deflate64 and rarfile deps for the same reasons, fewer deps to maintain and easier compatibility with `extract_archive`.

[the remark this answers is not in the document]
Comment (nilsleh) [reply requirements/datasets.txt:9]: I mean have to rewrite the whole data.py creation and file schema and downloading/extraction logic, which is always the most work when adding a dataset.

File: torchgeo/datasets/bigearthnet.py [hunk 6]
```diff
@@ -572,3 +582,444 @@ def plot(
...
+    def _extract(self) -> None:
+        """Extract the tarball parts.
+
+        For each modality (s1, s2, maps), its parts are concatenated together and then extracted.
+        """
+        chunk_size = 2**15  # same as used in torchvision and ssl4eo
+        for key, meta in self.metadata_locs.items():
+            if key == 'metadata':
+                continue
+            assert not os.path.isdir(os.listdir(self.root)[2])
```
Comment (nilsleh) [inline torchgeo/datasets/bigearthnet.py:929]: @adamjstewart any idea what could cause this error? Basically, when you check whether the tar ball file in the root is a directory it says no, but in the next line, when you do `os.path.join()` with the root, it all of the sudden is a directory... I don't see what I messed up.

Comment (adamjstewart) [reply torchgeo/datasets/bigearthnet.py:929]: Fixed in latest commit.

File: requirements/datasets.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply requirements/datasets.txt:9]: zstandard is no longer used now, right? We can remove all references to it.

File: torchgeo/datasets/bigearthnet.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/datasets/bigearthnet.py:807]: It looks like the raw data is uint16? int32 doesn't make much sense, no reason to cast dtype multiple times.

Comment (nilsleh) [reply torchgeo/datasets/bigearthnet.py:801]: I don't think pytorch has uint16 datatype, so I am guessing that is why it is being read as int32 so that the tensor can be built and then cast to float32

File: torchgeo/datasets/bigearthnet.py [hunk 7]
```diff
@@ -572,3 +576,430 @@ def plot(
...
+        523: 18,  # Sea and ocean
+    }
+
+    valid_splits = ('train', 'val', 'test')
+
+    def __init__(
+        self,
+        root: Path = 'data',
+        split: str = 'train',
+        bands: str = 'all',
```
Comment (adamjstewart) [inline torchgeo/datasets/bigearthnet.py:688]: People have complained that some datasets have (B1, B2, etc.) and others have s1, s2, etc. I know you're just matching the existing dataset, but should we change both to the former?

File: torchgeo/datasets/bigearthnet.py [hunk 8]
```diff
@@ -572,3 +576,430 @@ def plot(
...
+        exists = []
+        for key, metadata in self.metadata_locs.items():
+            if key != 'metadata':
+                exists.append(
+                    os.path.exists(os.path.join(self.root, self.dir_file_names[key]))
+                )
+            else:
+                exists.append(
+                    os.path.exists(os.path.join(self.root, self.dir_file_names[key]))
+                )
```
Comment (adamjstewart) [inline torchgeo/datasets/bigearthnet.py:864]: Maybe I'm blind, but what is the difference here?

File: torchgeo/datasets/bigearthnet.py [hunk 9]
```diff
@@ -572,3 +576,430 @@ def plot(
...
+                    exists.append(True)
+                else:
+                    exists.append(False)
+            else:
+                for fname in metadata['files']:
+                    fpath = os.path.join(self.root, fname)
+                    if os.path.exists(fpath):
+                        exists.append(True)
+                    else:
+                        exists.append(False)
```
Comment (adamjstewart) [inline torchgeo/datasets/bigearthnet.py:883]:
```suggestion
                    exists.append(os.path.exists(fpath))
```
Same above

**Review by adamjstewart — APPROVED** [review]
Minor comments, can probably merge if you're in a rush.

**Review by adamjstewart — APPROVED** [review]
I'm happy if you're happy.
